zeroex/order_validator.go

```diff
@@ -47,16 +41,105 @@ type OrderInfo struct {
 	TxHash common.Hash
 }
 
-// SuccinctOrderInfo represents the necessary information about an order without including
-// the order itself
-type SuccinctOrderInfo struct {
+// RejectedOrderInfo encapsulates all the needed information to understand _why_ a 0x order
+// was rejected (i.e. did not pass) order validation. Since there are many potential reasons, some
+// Mesh-specific, others 0x-specific and others due to external factors (i.e., network
+// disruptions, etc...), we categorize them into `Kind`s and uniquely identify the reasons for
+// machines with a `Code`
+type RejectedOrderInfo struct {
+	OrderHash   common.Hash
+	SignedOrder *SignedOrder
+	Message     string
+	Kind        RejectedOrderKind
+	Code        RejectedOrderCode
+}
+
+// AcceptedOrderInfo represents an fillable order and how much it could be filled for
+type AcceptedOrderInfo struct {
 	OrderHash                common.Hash
+	SignedOrder              *SignedOrder
 	FillableTakerAssetAmount *big.Int
-	OrderStatus              OrderStatus
 }
 
-// OrderHashToSuccinctOrderInfo maps orderHashes to their corresponding succinctOrderInfo
-type OrderHashToSuccinctOrderInfo map[common.Hash]*SuccinctOrderInfo
+// RejectedOrderCode enumerates all the unique reasons for an orders rejection
+type RejectedOrderCode uint8
+
+// RejectedOrderCode values
+const (
+	ROInvalid RejectedOrderCode = iota // Not an actual RejectedOrderCode. Occupies default zero-value.
+	RORequestFailed
+	ROInvalidMakerAssetAmount
+	ROInvalidTakerAssetAmount
+	ROExpired
+	ROFullyFilled
+	ROCancelled
+	ROSignatureInvalid
+	ROInvalidMakerAssetData
+	ROInvalidTakerAssetData
+	// ROUnfunded is a catch-all for when either the maker or taker have insufficient
+	// balance or allowance set to fullfil the order.
+	ROUnfunded
+)
+
+// ConvertRejectOrderCodeToOrderEventKind converts an RejectOrderCode to an OrderEventKind type
+func ConvertRejectOrderCodeToOrderEventKind(rejectedOrderCode RejectedOrderCode) (OrderEventKind, bool) {
+	switch rejectedOrderCode {
+	case ROExpired:
+		return EKOrderExpired, true
+	case ROFullyFilled:
+		return EKOrderFullyFilled, true
+	case ROCancelled:
+		return EKOrderCancelled, true
+	case ROUnfunded:
+		return EKOrderBecameUnfunded, true
+	default:
+		// Catch-all returns Invalid OrderEventKind
+		return EKInvalid, false
+	}
+}
+
+// GetMessageForRejectOrderCode returns the corresponding message for the RejectedOrderCode
+func GetMessageForRejectOrderCode(rejectedOrderCode RejectedOrderCode) string {
```

Comment:
Maybe instead of using this helper function we could add the message itself to the `RejectedOrderCode` type?

Something like:

```go
type RejectedOrderCode struct {
    Code string
    Message string
}

var (
    RORequestFailed RejectedOrderCode = {
        Code: "RequestFailed"
        Message: "network request to Ethereum RPC endpoint failed"
    },
    // ...
)
```

This has two benefits: (1) It puts the definition of the code and its message closer together and (2) it makes it nearly impossible to forget to associate a code with a message.

Maybe if we go with this approach we also rename the `RejectedOrderCode` type? Something like `RejectedOrderStatus`?